rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,404 @@
+> to finalize the upload. The endpoint will look for the expected file, and if it's found, update the database with the file
+> info.
+
+> TODO: How do we log when a download fails, and how is that propagated to the user.
+
+#### Local to Cloud Flow
+
+Local to cloud upload will allow users to upload files from their local machine to the cloud. Uploads happen though the
+DP browser App directly to S3. For small files <100MB, uploads can be accomplished using pre-signed URLs. For large files >5GB
+a multipart upload must be performed. Now to walk through the local to cloud flow, see figure 4.
+
+![Local to Cloud](https://app.lucidchart.com/publicSegments/view/80361b71-5317-478b-921e-0ead5152c865/image.png)
+**Figure 4:** Architecture for uploading from a local machine to our cloud.
+
+##### 1. Upload File
+
+Using the browser App, a file is selected for upload and a request is sent to the DP backend upload the file.
+
+##### 2. Request Token or Presigned URL
+
+Depending on the size of the file a different response will be returned. If an access token is returned, a
+multipart upload is started using the AWS SDK. If a pre-signed URL is returned an upload is started using the URL.
+
+##### 3. Upload Directly S3
+
+The implementation uses an access token and the AWS JS SDK to upload the file directly to S3. The hash of each chunk must be  
+calculated locally and supplied in the [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax)
+as part of Put Object request. This will verify the integrity of each chunk during the upload. If the hash does not match
+the chunk will be rejected. The AWS S3 API will do the heavy lifting as far as managing the [multipart upload](https://docs.aws.amazon.com/AmazonS3/latest/dev/uploadobjusingmpu.html). 
+The access token will provide sufficient permissions to allow a multipart upload using AWS JS SDK.
+
+##### User Upload Limitation
+
+S3 is the CZI CSP where uploaded datasets are stored. Since S3 allows small files to be uploaded without using multipart
+upload, our implementation can generate a single pre-signed URL when securely uploading small data sets.
+
+For large files, AWS S3 requires that the files be uploaded using their multipart upload process. This requires multiple
+requests to S3, and cannot be completed with a single pre-signed URL. In this case, a temporary security token is generated.
+The token grants the bearer directs access to the S3 bucket. This allows the browser application to use
+the AWS JS SDK to upload a large file in multiple parts with ease. It is important to limit the permissions of the token
+to only allow it to upload the expected file. There is also a maximum time to live of one hour for the token, therefore if
+the upload is not complete before then, a new token must be issued.
+
+##### Cleanup Multipart Uploads
```

Comment:
note we have support for this [here](https://github.com/chanzuckerberg/cztack/tree/main/aws-s3-private-bucket)

Reply:
add in the reference